### ISSUE TITLE
Subscriptions Empty - access to /feeds/videos.xml 404s

### ISSUE THREAD

**Issue by taleteller** [anxdpanic/plugin.video.youtube#566]
### Context

Please provide any relevant information about your setup

* Add-on Version: 7.0.2.2-1
* Kodi Version: 20.2
* Kodi GUI Language: German
* Region: Germany

------

### Expected Behavior

Subscriptions should load

------

### Current Behavior

This started to happen very recently and I cannot say if its another change to API or just a regional outtake.

When opening the "my subscriptions" page it comes up empty. looking into the logs I noticed the same error coming up over and over again:

```
2024-01-29 18:05:36.481 T:3321    error <general>: requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://www.youtube.com/feeds/videos.xml?channel_id=UCchBatdUMZoMfJ3rIzgV84g
```
A 404 sounds not good, and if i enter the url into my browser i get an identical error stating:

```
The requested URL /feeds/videos.xml?channel_id=UCchBatdUMZoMfJ3rIzgV84g was not found on this server. That’s all we know.
```

Well OK, probably the channel is down or something? No because the channel itself using its ID comes up fine. 

```
https://www.youtube.com/channel/UCchBatdUMZoMfJ3rIzgV84g
```

Visiting the same channel via Subscriptions -> Channel from the plugin it comes up fine and lists the videos made by this specific channel. However it seems to use a different approach:

``` 
2024-01-29 18:28:51.588 T:3071     info <general>: [plugin.video.youtube] Running: YouTube (7.0.2.2+unofficial.1) on N***** (Kodi-20.2) with Python 3.11.4
                                                   	Path: /subscriptions/list/
                                                   	Params: {'page': '3', 'page_token': 'CDwQAA'}
2024-01-29 18:28:52.716 T:3071     info <general>: [plugin.video.youtube] Running: YouTube (7.0.2.2+unofficial.1) on N***** (Kodi-20.2) with Python 3.11.4
                                                   	Path: /channel/UCchBatdUMZoMfJ3rIzgV84g/
                                                   	Params: {}
```

**Comment by benyamin-codez** [comment]
Same issue in English in APAC region on same Kodi and Add-on version...

Random channels are 404 via `/feeds/videos.xml?channel_id=` it would seem for the last several hours. One instance of all channels being 404, but back to seemingly random 404s, yet still many to most channels being 404.

**Comment by misslisa62tx** [comment]
Same issue with 7.0.3+beta.1.  Started yesterday evening or this morning.  I get video lists in my subscriptions, but they are old and incomplete, not what should be at the top of the list.  404 errors in the log,  I've tried clearing the data cache and rebooting and logging out and back in with no luck.

**Comment by rwlove** [comment]
Same with:
CoreELEC 20.3-Nexus
YouTube 7.0.2.2

**Comment by kedarz** [comment]
Same issue with 7.0.3+beta.1. and 7.0.3+beta.2
Nexus 20.3

**Comment by toupeira** [comment]
Some related discussions on Hacker News:

- [Tell HN: YouTube RSS feeds are gone](https://news.ycombinator.com/item?id=39179446)
- [Is YouTube starting to protect channel RSS feeds?](https://news.ycombinator.com/item?id=39175495)

No conclusion so far, this could be a temporary bug or a deliberate change.

**Comment by taleteller** [comment]
> No conclusion so far, this could be a temporary bug or a deliberate change.

Interesting, from our experience I would say the disabled it deliberately as a trial balloon. If there is no big backlash they keep it to hinder alternate players.

**Comment by neo-neo1** [comment]
Would be a gigantic blow to Kodi if we can't even watch YouTube anymore... Might as well switch to Android and Smart Tube App. But I prefer Kodi.

**Comment by taleteller** [comment]
Watching the Videos works fine, listing the videos for a single channel works fine as well. Just the subscription overview seem to be broken once more.

**Comment by neo-neo1** [comment]
> Watching the Videos works fine, listing the videos for a single channel works fine as well. Just the subscription overview seem to be broken once more.

Without My Subscriptions working this kills the use case of YouTube for majority of us.

**Comment by taleteller** [comment]
> Watching the Videos works fine, listing the videos for a single channel works
> fine as well. Just the subscription overview seem to be broken once more.
> Without My Subscriptions working this kills the use case of YouTube for majority
> of us.

My point is, the "My Subscriptions" page is combining the rss feeds of all your
subscribed channels while the "Channels" page is using a different approach. Therefore
there is a possible solution adopting the second way for the Subscriptions page as well.

**Comment by neo-neo1** [comment]
> Watching the Videos works fine, listing the videos for a single channel works fine as well. Just the subscription overview seem to be broken once more. Without My Subscriptions working this kills the use case of YouTube for majority of us.
> My point is, the "My Subscriptions" page is combining the rss feeds of all your subscribed channels while the "Channels" page is using a different approach. Therefore there is a possible solution adopting the second way for the Subscriptions page as well.

The reason this "different approach" was implemented in the first place is YouTube API ended access to My Subscription Feed a few years ago. If you used this Addon for a long time, you'd know. Three was a switch.

**Comment by taleteller** [comment]
> The reason this "different approach" was implemented is YouTube API ended
> access to My Subscription Feed a few years ago. If you used this Addon for
> a long time, you'd know.

Yes but thats still not my point, it does not matter if you combine the rss feeds
or api video lists for all channels to get the same result.

**Comment by benyamin-codez** [comment]
The feed provides video metadata in XML format which is parsed to provide the view in Kodi. If removed, then the only alternative would be to scrape the HTML, which is prone to breaking with page changes.

This same issue happened about 4 months ago. It would seem there is a repeat performance, and perhaps the waters are being tested once more...

It is also noteworthy that the current "My Subs" mechanism won't list channels marked as Made ForKids, per #116. A better mechanism would be awesome, but the present API plus Feed combo is the lowest maintenance solution I am aware of...

**Comment by neo-neo1** [comment]
> The reason this "different approach" was implemented is YouTube API ended access to My Subscription Feed a few years ago. If you used this Addon for a long time, you'd know.
> Yes but thats still not my point, it does not matter if you combine the rss feeds or api video lists for all channels to get the same result.

There's no other way. If RSS doesn't work your forced to use a clunky HTML parser which will needed to be updated often.

**Comment by benyamin-codez** [comment]
> Yes but thats still not my point, it does not matter if you combine the rss feeds or api video lists for all channels to get the same result.

I see your point that the add-on gets a list via Subscriptions... I'll check how...

UPDATE:
Looks like it uses the `perform_v3_request` rather than a dedicated session threadpool...

**Comment by benyamin-codez** [comment]
Looks like master is well ahead of v7.0.2.2 for `resources/lib/youtube_plugin/youtube/client/youtube.py`...

Perhaps @MoojMidge has already fixed this...?

It looks like master is still using a dedicated connection pool and thus bypasses `perform_v3_request`...

**Comment by kedarz** [comment]
With the NewPipe is still working. There is  "What's new"

**Comment by misslisa62tx** [comment]
Mine just started working again.

**Comment by taleteller** [comment]
@benyamin-codez: not sure where you see that, `youtube.py` on master is 2 weeks old and this looks like the current way to me:

```
                for channel_id in sub_channel_ids:
                    thread = threading.Thread(
                        target=fetch_xml,
                        args=('https://www.youtube.com/feeds/videos.xml?channel_id=' + channel_id,
                              responses)
                    )
                    threads.append(thread)
                    thread.start()

                for thread in threads:
                    thread.join(30)
```

**Comment by neo-neo1** [comment]
Don't know if it's a coincidence but I upgraded to 7.0.3 Beta 1 and it started working.

And I suggest the upgrade anyways, lots of new UI updates. Thanks guys

**Comment by toupeira** [comment]
> Mine just started working again.

Confirmed here too on latest stable 7.0.2.2 :tada:

**Comment by neo-neo1** [comment]
**Mark this as Closed**

**Comment by benyamin-codez** [comment]
> @benyamin-codez: not sure where you see that, `youtube.py` on master is 2 weeks old and this looks like the current way to me...

Yep, I saw that too, hence my edit above. Looks like quite a bit more has been added too.

Per posts above, this appears to be resolved for now anyway.

**Comment by karlo2105** [comment]
I am on Youtube 7.0.3 beta 1 and My subscription doesn't work. Since today, i noticed playback history doesn't work either.

**Comment by taleteller** [comment]
@karlo2105 can you recheck that? for most (all?) of us it started working a short time ago, and I think nobody has the desire to reopen this ;-)

**Comment by karlo2105** [comment]
@taleteller

24-01-29 22:57:16.861 T:12225    info <general>: [plugin.video.youtube] Running: YouTube (7.0.3+beta.1) on Nexus (Kodi-20.3) with Python 3.11.2
                                                   Path: /special/new_uploaded_videos_tv/
                                                   Params: {'category_label': '[B]Mes abonnements[/B]'}
2024-01-29 22:57:17.195 T:12225   error <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: string indices must be integers, not 'str'
                                                   Traceback (most recent call last):
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/plugin.py", line 17, in <module>
                                                       runner.run(__provider__)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/kodion/runner.py", line 60, in run
                                                       __RUNNER__.run(provider, context)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/kodion/plugin/xbmc/xbmc_runner.py", line 59, in run
                                                       results = provider.navigate(context)
                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/kodion/abstract_provider.py", line 116, in navigate
                                                       result = method(context, re_match)
                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/kodion/register_provider_path.py", line 19, in wrapper
                                                       return func(*args, **kwargs)
                                                              ^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/provider.py", line 675, in _on_yt_specials
                                                       return yt_specials.process(category, self, context)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/yt_specials.py", line 327, in process
                                                       return _process_new_uploaded_videos_tv(provider, context)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/yt_specials.py", line 304, in _process_new_uploaded_videos_tv
                                                       return tv.my_subscriptions_to_items(provider,
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/tv.py", line 37, in my_subscriptions_to_items
                                                       channel = item['channel'].lower()
                                                                 ~~~~^^^^^^^^^^^
                                                   TypeError: string indices must be integers, not 'str'
                                                   -->End of Python script error report<--

**Comment by taleteller** [comment]
@karlo2105 thats issue #555 you probably need to clear you cache. I stumbeled over it searching an issue before opening this one.

**Comment by karlo2105** [comment]
@taleteller
After deleting data cache, my subscriptions works again. On the other side history playback doesn't display.

**Comment by MoojMidge** [comment]
> Perhaps @MoojMidge has already fixed this...?
> 
> It looks like master is still using a dedicated connection pool and thus bypasses `perform_v3_request`...

I'm in the process of migrating some channel related functions, primarily for playlists (saved, watch list, history list) to use the V1 innertube API, rather than the V3 Data API which no longer supports that functionality, or the V1 leanback TV API which no longer provides the same details that it used to.

Could do the same for the feeds, which I think is similar to what NewPipe does when fast mode is enabled. This is also similar to how My Subscriptions used to be fetched.

`perform_v3_request` has been replaced by `api_request` in master and uses a connection pool in the background. It is just not used for multiple parallel requests using threads at the moment, but there is nothing preventing it from being used like that except for API quota usage. This would be the most reliable method, but may be slow and result in API quota exhaustion for people with many subscriptions.

There is no seperate connection pool used anywhere now, and once all the V1 innertube API request changes are made, the next step will be to make the base request class able to handle multiple requests automatically using a thread pool or similar, without having to manually open threads.

**Comment by neo-neo1** [comment]
Edit: 7.0.3 beta 2 seems to have fixed the issues I had 


> I am on Youtube 7.0.3 beta 1 and "My subscription" gives an error. Since today, i noticed playback history doesn't work either.

Same problem here. "My Subscriptions" doesn't work on 7.0.3 beta 1 anymore. Clearing cache didn't fix anything. 

Had to uninstall 7.0.3 and install 7.2.2 to get things working. Tried installing 7.0.3 beta 1 on a fresh box (no -upgrade) and resulted in same problem. People on Kodi Forums reported the same problem with 7.0.3

**Comment by neo-neo1** [comment]
Aannnndd "My Subscriptions" just stoped working, again. Seems it's been in and out throughout the day.

**Comment by karlo2105** [comment]
I upgraded Youtube 7.0.3 beta 2 and i confirm "My subscriptions" tab is working. Unfortunately, to be able to play smoothly video with subtitles, i had to disable dash otherwise it is in slow motion. Any workaround for this instead of disabling dash?
Youtube 7.0.1 used to work till yesterday.

**Comment by benyamin-codez** [comment]
> I'm in the process of migrating some channel related functions, primarily for playlists (saved, watch list, history list) to use the V1 innertube API, rather than the V3 Data API which no longer supports that functionality, or the V1 leanback TV API which no longer provides the same details that it used to.
> 
> Could do the same for the feeds, which I think is similar to what NewPipe does when fast mode is enabled. This is also similar to how My Subscriptions used to be fetched.

Would requesting the channel list using either the innertube or the leanback TV APIs also fix issue #116...?  This is where channels marked as MadeForKids are not returned in the list, e.g. colinfurze. This affects Subscriptions too, not just My Subscriptions. You can tell on the YouTube channel page that a channel is affected because the notification feature is disabled. I presume the APIs you mention are REST APIs, so do provide some degree of maintainability. 

> `perform_v3_request` has been replaced by `api_request` in master and uses a connection pool in the background. It is just not used for multiple parallel requests using threads at the moment, but there is nothing preventing it from being used like that except for API quota usage. This would be the most reliable method, but may be slow and result in API quota exhaustion for people with many subscriptions.

Probably not applicable to quota usage, but I presume in your current implementation the `api_request` ConnectionPool is effectively only using one connection, as it is only accepting one request per instantiation...? I think I would probably come close to hitting the quota limit. Anyone with 250+ subscriptions could possibly hit it within a day. Anyone with 400+ subscriptions will most likely feel some pain... I agree though, it is probably most reliable. Would the abovementioned APIs have the same problem? Perhaps they are a better choice.

> There is no seperate connection pool used anywhere now, and once all the V1 innertube API request changes are made, the next step will be to make the base request class able to handle multiple requests automatically using a thread pool or similar, without having to manually open threads.

I think the challenge here is any decent thread pool would require the calling functions to be instantiated with an array or dict of job parameters so that the `ThreadPoolExecutor()` can process all the requests essentially as a **batch job** before shutting down the pool (per my comments in issue #529). This is probably a fundamental architectural change.  I good example of this is the changes I made in issue #529 to `fetch_xml()` with increasing complexity for various scenarios.  

Thanks @MoojMidge for taking the time to post your insights.

**Comment by MoojMidge** [comment]
@karlo2105 - if you have a problem create a new issue and provide some details (and a debug log).

**Comment by MoojMidge** [comment]
@benyamin-codez 
> Would requesting the channel list using either the innertube or the leanback TV APIs also fix issue #116...?

I think they are the same API, just providing different presentational data based on the UI of the client making the request. The TV specific leanback UI has been discontinued, but the same API is used by the other official Youtube clients and should fix #116.

Unfortunately using the V1 API is a bit of a moving target. Just started using it for related videos, but after these recent Youtube changes, the presentational data now no longer has the related video info, meaning it will need to be scraped from the main website with continuation, or from a TV client without continuation.

> Probably not applicable to quota usage, but I presume in your current implementation the `api_request` ConnectionPool is effectively only using one connection, as it is only accepting one request per instantiation...?

Not really. A session is created as a class attribute of the base request class. api_request is just a method of a sub-class of the base request class. For every plugin invocation, this session is used for all requests. Most listings typically involve multiple API requests, which are partially processed in parallel or loaded from cache where possible, but normally this is only noticeable on very large listings e.g. loading large playlists or multiple playlists or other combined listings. The related videos listing for example uses the api_request method and would fill the ConnectionPool.


> Would the abovementioned APIs have the same problem? Perhaps they are a better choice.

Maybe. They are not public APIs and are therefore subject to change and break without notice. Like what has just happened.

> I think the challenge here is any decent thread pool would require the calling functions to be instantiated with an array or dict of job parameters so that the `ThreadPoolExecutor()` can process all the requests essentially as a **batch job** before shutting down the pool (per my comments in issue #529). This is probably a fundamental architectural change.

That is the intent, but it doesn't require many more major changes. The called function, and the accompanying data structure that the function would mutate, can be specific to each required task. The idea would be to create a generic thread handler that can complete these bespoke tasks without needing to know anything about the called function or data. If the called function makes a request using the base request class or a subclass, then that will automatically utilise the existing connection pool.

A very basic implementation of this is in https://github.com/anxdpanic/plugin.video.youtube/blob/8c543b78a4a18d9f1ab28ce21321463b6482691b/resources/lib/youtube_plugin/youtube/helper/v3.py#L197C1-L331C40
In this example the number of threads is contrained by the type of data being processed (videos, channels, playlists, fanart, cache) so a simple loop can create the required threads then monitor thread state.

The common called function makes multiple requests, updates shared data, and fetches then updates the cache. The parallel requests are all handled automatically, as is the sqlite database access. This is all kept seperate from the called function, which is also kept seperate from the thread handling.

The main thing that needs to be improved is to make the thread handling more robust, while managing system resources and monitoring thread status. And doing so in a generic manner so that it can be used elsewhere in the plugin.


> I good example of this is the changes I made in issue #529 to `fetch_xml()` with increasing complexity for various scenarios.

I think the concept is sound, but my aim would be to completely de-couple the threaded tasks from the initiation/execution/monitoring/completion/closing of the threads.

**Comment by benyamin-codez** [comment]
@MoojMidge, you have been very busy...!  The example you gave was quite elegant.

>The main thing that needs to be improved is to make the thread handling more robust, while managing system resources and monitoring thread status. And doing so in a generic manner so that it can be used elsewhere in the plugin.
>...
>I think the concept is sound, but my aim would be to completely de-couple the threaded tasks from the initiation/execution/monitoring/completion/closing of the threads.

Decoupling threaded tasks from their management was exactly the challenge I faced in issue #529. You have the bare bones `Threading.Thread` on one hand and the fully managed `Concurrent.Futures` on the other. You may want to peek into the latter to see if any of the thread management mechanisms and techniques can be ported... Ultimately I went with `Futures` because it essentially resolved the specific issue in #529, i.e. threads failing to start. As to whether there's a specific mechanism as to why or if it is because the relevant code slowed things down a little, I couldn't say... 8^d
